[RFC] transport: spdm-storage: add support

twilfredo opened the pull request:
As per DMTF DSP0286 (pending ratification), this series adds supports to `spdm-utils` to communicate `spdm` with `scsi` and `nvme` devices over the respective security commands. This series also adds  support for a `qemu-server`. Which can to be connected to by `qemu`, and implement an SPDM responder for SCSI/ATA & NVMe devices.

twilfredo commented:
@alistair23

twilfredo commented:
todo:
1. DMTF DSP0286 Ratification
2. `libspdm` storage transport support upstream
3. `qemu` nvme responder socket support upstream

twilfredo commented:
@alistair23 this patch has README instructions for tcmu-runner, should we remove those?

alistair23 commented:
> Note: CI failure is for tock-responder seems unrelated to this

```

error[E0599]: no variant or associated item named `Storage` found for enum `TransportLayer` in the current scope
   --> /home/runner/work/spdm-utils/spdm-utils/src/libspdm/spdm.rs:511:25
    |
286 | pub enum TransportLayer {
    | ----------------------- variant or associated item `Storage` not found for this enum
...
511 |         TransportLayer::Storage => {
    |                         ^^^^^^^ variant or associated item not found in `TransportLayer`
```

Seems very related

alistair23 commented on Cargo.toml:
```diff
@@ -61,6 +62,7 @@ colored = { version = "2.1", optional = true }
 which = { version = "6.0", optional = true }
 asn1-rs = { version = "0.6", optional = true }
 x509-parser = { version = "0.15", optional = true }
+page_size = { version = "0.6.0", optional = true }
```
Can we just use "0.6", that way we get small updates

alistair23 commented:
> @alistair23 this patch has README instructions for tcmu-runner, should we remove those?

Yeah, I think so. Also if possible it might be worth splitting the NVMe and SCSI parts out. It'll help make the PR smaller

twilfredo commented:
> > Note: CI failure is for tock-responder seems unrelated to this
> 
> ```
> 
> error[E0599]: no variant or associated item named `Storage` found for enum `TransportLayer` in the current scope
>    --> /home/runner/work/spdm-utils/spdm-utils/src/libspdm/spdm.rs:511:25
>     |
> 286 | pub enum TransportLayer {
>     | ----------------------- variant or associated item `Storage` not found for this enum
> ...
> 511 |         TransportLayer::Storage => {
>     |                         ^^^^^^^ variant or associated item not found in `TransportLayer`
> ```
> 
> Seems very related

Yeah! I deleted that comment after I fixed it.. you got to it before haha my bad

twilfredo commented on README.md:
```diff
@@ -433,3 +448,78 @@ should show (ensure that INFO log level is enabled in SPDM-Utils).
+communicated to the NVMe device through the NVMe Security Send/Receive commands.
+
+As before, we need to start SPDM-Utils before QEMU, the following options are
+required.
```
Changed this PR to be NVMe only, so will add this for a different PR

twilfredo commented:
This PR now only adds the userspace side for NVMe only. SCSI and QEMU server to follow.

twilfredo commented on src/io_buffers.rs:
```diff
@@ -1,9 +1,86 @@
+enum BufferType {
+    MemAligned(Vec<u8>),
+    Default(Vec<u8>),
+}
```
fixed, now only a single type of aligned buffers.

alistair23 commented on src/io_buffers.rs:
```diff
@@ -1,44 +1,57 @@
-        Ok(buffers) => {
-            *(SEND_BUFFER.lock().unwrap()) = Some(buffers.0);
-            *(RECEIVE_BUFFER.lock().unwrap()) = Some(buffers.1);
+    if !send_recv_len.is_multiple_of(page_size::get()) {
```
Maybe a comment that this is due to libnvme

twilfredo commented:
@alistair23 ping

twilfredo commented:
@alistair23 ping

alistair23 commented on src/nvme.rs:
```diff
@@ -0,0 +1,679 @@
+    ///
+    ///  A mutable reference to @dptr is stored within the returned struct, thus,
+    ///  the underlying memory pointed to by @dptr must not be free(d) until
+    ///  after the data is transferred. Ex: after invoking `nvme_security_send()`.
```
This seems very un-Rust like. We should be able to use ownership to make sure you can't drop the memory. (at least not without `unsafe`)

twilfredo replied:
hmmm... I think we would have to do something like then, because the data field will just take a pointer. So to take ownership of the data we would need some wrapper like this, where `args.data` points to `data`.

```
pub struct NvmeSecuritySendCommand {
    data: Vec<u8>,
    args: nvme_security_send_args,
}
```

Do you have any better ideas?

alistair23 replied:
You could also store it in `NvmeSecSendCmds`. Then have a `NvmeSecSendCmds::new()` to create the struct and a `get_args()` function to get the args to pass to libnvme

twilfredo replied:
fixed now!

alistair23 commented on src/nvme.rs:
```diff
@@ -0,0 +1,701 @@
+        };
+
+        let mut transport_message_len = message_size;
+        let mut transport_message = ptr::null_mut();
```
Why is this `NULL`? Especially when the size isn't 0

twilfredo replied:
libspdm sets the `transport_message` by walking back the number of transport header bytes from `message`. But before doing that it needs to assert the length (so that we don't walk back into invalid memory).  So `transport_message_len` cannot be zero currently. See below (transport_message_len is transport_message_size in libspdm.

```
    if (*transport_message_size < message_size + sizeof(libspdm_storage_transport_virtual_header_t)) {
        *transport_message_size = message_size + sizeof(libspdm_storage_transport_virtual_header_t);
        return LIBSPDM_STATUS_BUFFER_TOO_SMALL;
    }
```
This function exists to extract the storage header that is already bound to the message so we can setup the storage API calls to the device. This is a bit clunky, so we could look at improving this api later.
